### ISSUE TITLE
[MIR] Fuse the bounds check into one unsigned compare and eliminate monotone-induction in-loop checks

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1081]
Baselines: Align `8c8bfbc7a3169e84ecc8415f5149ab8c61afe863` (alignc 0.7.5, LLVM 22.1.8), align-llm
`b72b0b48`, Apple M1, `--profile release`, default `--target-cpu baseline`.

## Problem

Align emits a bounds check as **two signed compares plus an `or`**, and a range check as **five
operations**. The `idx < 0` half is dead by construction (a length is non-negative), and for a loop indexed
by a monotone induction variable the whole check is loop-invariant in the sense that it can be proved once
in the preheader. Neither happens today.

The consequence is not just the instruction count: **two or more surviving in-loop bounds checks switch
LLVM's LoopVectorizer off entirely.** The threshold is exactly 1, measured.

This is part 3 of a single **loop-facts contract** (see *Relation*): on a real kernel, fixing the check
shape *alone* produces zero vector instructions, and so does fixing the alias facts alone — only the
conjunction vectorizes.

### 1. The emitted check shape

`crates/align_mir/src/lib.rs:11876-11913` — `emit_bounds_check`, two signed compares and an `or`:

```rust
fn emit_bounds_check(b: &mut Builder, idx: &Operand, len: Operand) -> BlockId {
    let lo = b.fresh_value(Ty::Bool);
    b.push(Stmt::Let(
        lo,
        Rvalue::Bin(
            BinOp::Lt,
            idx.clone(),
            Operand::Const(Const::Int(0, i64_ty())),
        ),
    ));
    let hi = b.fresh_value(Ty::Bool);
    b.push(Stmt::Let(
        hi,
        Rvalue::Bin(BinOp::Ge, idx.clone(), len.clone()),
    ));
    let oob = b.fresh_value(Ty::Bool);
    b.push(Stmt::Let(
        oob,
        Rvalue::Bin(BinOp::Or, Operand::Value(lo), Operand::Value(hi)),
    ));
```

`emit_range_bounds_check` (`:12591`) is the same idea with five operations (`start < 0`, `start > end`,
`end > len`, two `or`s), and `emit_vec_bounds_check` (`:12576`) delegates to it. There are 13 call sites of
`emit_bounds_check`. `git grep -n 'Profile|opt_level' -- crates/align_mir/src` finds **no profile
plumbing**: the check is identical in `dev`, `release` and `fast`, and there is no unchecked escape hatch.

### 2. The count, not the presence, is the vectorization gate

Same loop shape, different index provenance, `--profile release`:

| kernel | surviving in-loop checks | `default<O2>` result |
|---|---|---|
| `i < xs.len()`, read `xs[i]` | 0 | vectorized `<4 x float>` |
| `i < b.len()`, read `b.u8(i)` (`slice<u8>`) | 0 | vectorized `<16 x i8>` |
| read `src[i]`, write `dst[i]` (separate lengths) | 1 | loop-versioned: vector prefix + scalar remainder |
| `w[r*d+c]` and `x[c]`, bound `c < d` | 2 (+1 invariant `base<0`) | **not vectorized at all** |

The matvec inner loop is **9 instructions and 3 branches per single f32 multiply-accumulate**:

```
20: tbnz x10,#0x3f, <trap>   ; base<0 — LOOP-INVARIANT, not hoisted
28: b.ge <trap>              ; base+c >= w.len()
2c: cbz  x11, <trap>         ; x.len() exhausted
30: ldr s1,[x0,x9,lsl #2]  34: ldr s2,[x2],#0x4
44: fmul s1,s1,s2          48: fadd s0,s0,s1
```

Pre-slicing the row in source (`wr := w[r*d .. r*d+d]`) moves one guard to the preheader, leaving **one** —
and LLVM then vectorizes to `<4 x float>`. That is the proof that the count is the gate.

### 3. Even the *explicit* `vecN` load loop pays for it

`xs.load(i)` into a `vec4<f32>` is the most explicit SIMD spelling the language has. Emitted machine code
of its inner loop, `--target-cpu baseline`:

```
30: mov  x11, #0x7ffffffffffffffc   ; hoisted constant
<L>:
34: cmp  x8, x11
38: add  x1, x8, #0x4
3c: b.eq <range trap>           ; `start > end` overflow test — LOOP-INVARIANT, not hoisted
40: cmp  x1, x2
44: b.gt <range trap>           ; `end > len`
48: ldr  x8, [x0]               ; header PTR reloaded
4c: cmp  x1, x9
50: ldr  q1, [x8, x10]
54: add  x10, x10, #0x10
58: mov  x8, x1
5c: fadd.4s v0, v0, v1
60: b.lt <L>
```

**12 instructions and 3 branches per 4 floats**, where `ldr q1` + `fadd.4s` + the back-edge would do. Two
of the twelve are a loop-invariant `idx == i64::MAX-3` overflow guard that was never hoisted.

### 4. On real align-llm code

| fact | count |
|---|---|
| `align_rt_bounds_fail` call sites surviving in the 746,869-line whole-program optimized IR | **2,311** |
| `align_rt_range_fail` call sites, same IR | **790** |
| trap branches (`bounds_fail` + `range_fail` + `utf8_boundary_fail`) as a share of `br i1` in Align function bodies, drop glue excluded | 3,605 of 29,726 = **12.1 %** |
| `tokenizer_qwen2$encode_raw_append` trap branches | 159 of 350 = **45 %** |
| unfolded `tbnz/tbz xN, #0x3f` sign-bit tests in Align code | **1,358** across 409 functions |
| `vector.body` in 2,108 defined Align functions | **73** |

Top owners of the surviving sign tests: `expert_trace$scan` (60), `residency_sim$simulate_mode` (32),
`residency_sim$replay` (31), `runtime_generation$application_capacity` (31),
**`moe_decode_step$run_moe_layer` (27)**, **`tokenizer_qwen2$encode_raw_append` (27)**,
`moe_layer_forward$scan_transcript_after` (24), `alignpack$plan_layout` (20). Some are loop-invariant and
un-hoisted: `alignpack$decode_pack` L15 carries **9** inside a 320-instruction loop body.

LLVM reports **311** *"failed to hoist load with loop-invariant address because load is conditionally
executed"* remarks, concentrated on the two-slice-index shape: `layer_forward.align:1329:5` (62),
`model_forward.align:1175:5` (31), `:1165:5` (31), `moe_layer_forward.align:1596:5` (30),
`layer_olmoe.align:1339:9` (15). The last is the textbook case, `align-llm/src/layer_olmoe.align:1339`:

```align
        if ids[check_token * n_expert_used + left] == ids[check_token * n_expert_used + right] {
```

The hottest instance is `align-llm/src/decode_step.align:636-660` (`stage_past_k`), whose inlined inner
loop pays **two bounds branches and two header reloads per 4 bytes moved**:

```
10005f4f0: ldr x2, [x24, #0x8]      ; slice LEN reloaded
10005f4f4: cmp x19, x28
10005f4f8: add x1, x19, #0x4
10005f4fc: b.hi <bounds trap>       ; check 1
10005f500: cmp x1, x2
10005f504: b.gt <bounds trap>       ; check 2
10005f508: ldr x8, [x24]            ; slice PTR reloaded
10005f518: ldr w1, [x8, x19]
10005f51c: bl  <_align_rt_buffer_put>
```

### 5. This must land with the alias-facts change — measured conjunction

On the byte→typed-`f32` route kernel, hoisting the in-loop checks **alone** produces zero vector
instructions and no measurable change (59.4 µs/call on 152,064 elements), and so does adding
`noalias` to the `borrow` header alone; only both together give a `<4 x float>` main loop at
12.6 µs/call (4.7×). The four-variant table is in #1079 §5 and is not repeated here.

## Root cause

`emit_bounds_check`/`emit_range_bounds_check` construct the check from generic signed `BinOp`s with no use
of the length's non-negativity and no awareness of the enclosing loop's induction structure. There is no
MIR pass between lowering and codegen that recognizes a monotone index and moves the proof to the
preheader; the only existing elimination is the narrow byte-range proof
(`crates/align_mir/src/byte_ranges.rs`, hooked at `crates/align_mir/src/lib.rs:3419`).

## Proposed solution

One canonical check shape, plus one general elimination rule. Cheapest first; (1) is unconditionally
correct and independent.

**(1) Fuse to a single unsigned compare.** A length is non-negative by construction, so
`(idx < 0) || (idx >= len)` ≡ `(u64)idx >= (u64)len`. This halves the raw check, deletes the `or`, and
hands SCEV one monotone `icmp ult` instead of a disjunction. Zero semantic change; the same rewrite
applies to the three-compare `emit_range_bounds_check`, whose `start < 0` and `end > len` fuse the same
way. Applies to `emit_vec_bounds_check` by construction.

**(2) A general monotone-induction bounds-check elimination in MIR**, added beside `byte_ranges::simplify`
at the existing hook `crates/align_mir/src/lib.rs:3419`. For a loop exiting on `i >= N` with `i` stepped by
a positive constant and an index `a*i + b` with loop-invariant `a > 0` and `b`, emit **one preheader
guard** and drop the in-loop one. The guard is *moved*, not deleted, so the exact `(index, len)` diagnostic
and the first-failing-access behaviour survive. This is the general rule
`docs/impl/63-codegen-performance-audit.md:232` explicitly says is still missing
(*"This is not complete bounds-check elimination"*), stated for induction variables rather than for byte
ranges only.

**(3) Extend the same rule to stores through an `out`/`mut` slice parameter.** Today
`byte_ranges::effects_safe_until` (`crates/align_mir/src/byte_ranges.rs:414`) whitelists only reads plus
pushes to a *locally created* `array_builder` (`fresh()`, `:383`), so the realistic shape — read bytes,
write an out slice — keeps its guard. Align's exclusivity rules already make this sound, and without it
(2) does not reach the kernels that matter.

Deliberately **not** proposed: `llvm.assume`, loop-metadata overrides, or any profile-dependent check
removal. `docs/open-questions.md:4187-4189` rejects the first two as a general policy; (1)–(3) need
neither.

## Acceptance criteria

- `tbnz/tbz xN, #0x3f` count in Align code drops from 1,358 to < 100 (the genuinely user-written sign
  tests); `align_rt_bounds_fail` *call-site* count is unchanged by (1) — the check is fused, not removed —
  and the `(index, len)` diagnostic text is byte-identical.
- The matvec kernel `w[r*d+c] * x[c]` vectorizes to `<4 x float>` with no source change (today: 9
  instructions, 3 branches per MAC, not vectorized).
- The explicit `vec4` `xs.load(i)` accumulation loop's inner body drops from 12 instructions and 3
  branches to ≤ 5 instructions and 1 branch per 4 floats; the loop-invariant `idx == i64::MAX-3` guard is
  in the preheader.
- `stage_past_k`'s inlined loop emits 0 in-loop bounds branches.
- A zero-length slice, a length-1 slice, and an out-of-range first access each still trap at the same
  iteration with the same message (owner tests for each of (1), (2), (3)).
- Whole-program surviving trap call sites in align-llm drop materially from 2,311 + 790, and
  `vector.body` count rises from 73.
- With the alias-facts change also present, `bytes_to_f32_out` reaches a `<4 x float>` main loop; pin the
  conjunction so a future regression in either half is caught.

## Relation to existing issues/plans

Part 3 of the **loop-facts contract**, with:

1. *"[Codegen] Materialize a borrowed slice/array view header once and state its alias facts"* (#1079);
2. *"[Codegen] Attach `!range` to slice/array length loads so a length is known non-negative"*
   (#1080) — which makes (1) above provably correct to LLVM as well as to us;
3. **this issue**.

§5 is the reason they are one unit of work rather than three patches.

- **Partial prior coverage.** A restricted byte-range proof already ships
  (`docs/impl/63-codegen-performance-audit.md:199-240`, `:468-498`, measured 10.883 → 1.778 µs), extended
  by `docs/impl/64-composed-byte-optimization-plan.md`. `63-...md:232` states it is *"not complete
  bounds-check elimination"*. `docs/impl/62-decode-optimization-plan.md:47` argues LICM/BCE is *"already
  handled by stock LLVM"* on its fixed-40-element witness and asks to *"obtain the real residual loop
  before adding a pass"* — §2, §3 and §4 above are that residual loop.
- Mechanism behind the scalar loops in **#1064**; the comment on #1064 records the measurement.
- Not a duplicate of #1047, #1063, #1065 or #1066.

## Impact framing

In a real Metal generation profile Align-compiled code is ~0.1 % of CPU during GPU decode, so this is not
a headline tokens/s number today. Its scope is the CPU-side kernels Align actually owns: greedy logits
selection at **242 µs/token** over 151,936 logits and the sampler at **267 µs/token** over 152,064
(this host); the tokenizer path, which is **74.5 %
Align code** and where `encode_raw_append` spends 45 % of its branches on traps; the KV-plane staging and
verification paths; and the future in-language CPU inference path. Separately, 1,358 provably-dead sign
tests and 73 vector bodies across 2,108 functions in a 5.3 MB numeric binary is a code-quality statement
for the general "Align must emit fast binaries" goal.
